pkg/nsx/services/securitypolicy/builder.go

```diff
@@ -1888,6 +1882,51 @@ func (service *SecurityPolicyService) buildRulePortsString(ports *[]v1alpha1.Sec
 	return util.GenerateTruncName(common.MaxNameLength, portsString, "", suffix, "", "")
 }
 
+func (service *SecurityPolicyService) buildRulePortNumberString(port *v1alpha1.SecurityPolicyPort, hasNamedport bool, portNumber int) string {
+	// Build the rule port number string name for non named port.
+	// This is a common case where the string is built from port definition. For instance,
+	// - protocol: TCP
+	//   port: 8282
+	//   endPort: 8286
+	// The built port number string is: 8282.8286
+	// - protocol: UDP
+	//   port: 3308
+	// The built port number string is: 3308
+	if !hasNamedport {
+		if port.EndPort != 0 {
+			return fmt.Sprintf("%s.%d", (port.Port).String(), port.EndPort)
+		}
+		return fmt.Sprintf("%s", (port.Port).String())
+	} else {
+		// Build the rule port number string name for named port.
+		// The port number string is built from specific port number resolved from named port.
+		return fmt.Sprintf("%d", portNumber)
+	}
+}
+
+func (service *SecurityPolicyService) buildRulePortsNumberString(ports *[]v1alpha1.SecurityPolicyPort) string {
+	portsNumString := ""
+	if ports == nil || len(*ports) == 0 {
+		portsNumString = common.RuleAnyPorts
+	} else {
+		for idx, p := range *ports {
+			port := p
+			portNumString := service.buildRulePortNumberString(&port, false, -1)
+			if idx == 0 {
+				portsNumString = portNumString
+			} else {
+				portsNumString = strings.Join([]string{portsNumString, portNumString}, common.ConnectorUnderline)
+			}
+		}
+	}
+	return portsNumString
+}
+
+func (service *SecurityPolicyService) buildRuleHashString(rule *v1alpha1.SecurityPolicyRule) string {
+	serializedBytes, _ := json.Marshal(rule)
+	return fmt.Sprintf("%s", util.Sha1(string(serializedBytes)))
```

Comment:
Having checked function `util.Sha1`, the return string is with length of 40 chars, do we need to keep the complete string here as the unique part? Or can we truncate it with less chars?